R/pkg/R/DataFrame.R

```diff
@@ -1452,15 +1452,147 @@ setMethod("join",
+              joinType <- "outer"
+            } else if (all.x) {
+              joinType <- "left_outer"
+            } else if (all.y) {
+              joinType <- "right_outer"
+            }
+
+            # join expression is based on by.x, by.y if both by.x and by.y are not missing
+            # and by if by.x or by.y are missing or have different lengths
+            if (length(by.x) > 0 & length(by.x) == length(by.y)) {
+              joinX <- by.x
+              joinY <- by.y
+            } else if (length(by) > 0) {
+              joinX <- by
+              joinY <- by
+            } else {
+              stop("The intersection of dataframes is empty")
+            }
+
+            # sets alias for making colnames unique in dataframes 'x' and 'y'
+            colsX <- generateAliasesForIntersectedCols(x, names(x), by, suffixes[1])
+            colsY <- generateAliasesForIntersectedCols(y, names(y), by, suffixes[2])
+
+            # selects columns with their aliases from dataframes
+            # in case same column names are present in both data frames
+            xsel <- select(x, colsX)
+            ysel <- select(y, colsY)
+
+            joinColumns <- lapply(seq_len(length(joinX)), function(i) {
+              colX <- joinX[[i]]
+              colY <- joinY[[i]]
 
+              if (colX %in% by) {
+                colX <- paste(colX, suffixes[1], sep = "")
+              }
+              if (colY %in% by) {
+                colY <- paste(colY, suffixes[2], sep = "")
+              }
+
+              colX <- getColumn(xsel, colX)
+              colY <- getColumn(ysel, colY)
+
+              colX == colY
+            })
+
+            # concatanates join columns together with '&'
+            for (i in 1:length(joinColumns)) {
+              if (i == 1) {
+                joinExpr <- joinColumns[[i]]
+              } else {
+                joinExpr <- joinExpr & joinColumns[[i]]
+              }
+            }
+
+            joinRes <- join(xsel, ysel, joinExpr, joinType)
+
+            # sorts the result by 'by' columns if sort = TRUE
+            if (sort & length(by) > 0) {
+              colNameWithSuffix <- paste(by, suffixes[2], sep = "")
+              joinRes <- do.call("arrange", c(joinRes, colNameWithSuffix, decreasing = FALSE))
+            }
+
+            joinRes
+          })
+
+#' 
+#' Creates a list of columns by replacing the intersected ones with aliases.
+#' The name of the alias column is formed by concatanating the original column name and a suffix.
+#'
+#' @param x a DataFrame on which the
+#' @param allColNames a list of column names
+#' @param intersectedColNames a list of intersected column names
+#' @param suffix a suffix for the column name
+#' @return list of columns
+#'
+generateAliasesForIntersectedCols <- function (x, allColNames, intersectedColNames, suffix) {
+  # sets alias for making colnames unique in dataframe 'x'
+  cols <- lapply(seq_len(length(allColNames)), function(i) {
```

Comment:
I'm a bit confused - why create a seq of indices? instead can't we use lapply on allColNames?

Also, nit, why pass allColNames if we could just do `names(x)` here?